File: tools/codex-scheduler/app/job_runner.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path

from core import extract_assistant_answer
from task_store import update_task_file, utc_now


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Run Codex, persist JSONL, and optionally save its clean answer")
    parser.add_argument("--log", required=True, type=Path)
    parser.add_argument("--answer", type=Path)
    parser.add_argument("--task-state", type=Path)
    parser.add_argument("command", nargs=argparse.REMAINDER)
    args = parser.parse_args()
    if args.command and args.command[0] == "--":
        args.command = args.command[1:]
    if not args.command:
        parser.error("a command is required after --")
    return args


def task_cancel_requested(path: Path | None) -> bool:
    if path is None:
        return False
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return False
    return bool(data.get("cancel_requested")) if isinstance(data, dict) else False


def patch_task(path: Path | None, **values: object) -> None:
    if path is not None:
        update_task_file(path, values)


def main() -> int:
    args = parse_args()
    args.log.parent.mkdir(parents=True, exist_ok=True)
    if args.answer:
        args.answer.parent.mkdir(parents=True, exist_ok=True)

    lines: list[str] = []
    try:
        with args.log.open("w", encoding="utf-8", buffering=1) as log:
            proc = subprocess.Popen(
                args.command,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
                bufsize=1,
                start_new_session=True,
            )
            patch_task(
                args.task_state,
                status="running",
                pid=proc.pid,
                started_at=utc_now(),
                finished_at=None,
                error=None,
            )
            assert proc.stdout is not None
            for line in proc.stdout:
                log.write(line)
                lines.append(line)
            rc = proc.wait()
    except Exception as exc:
        patch_task(
            args.task_state,
            status="failed",
            pid=None,
            finished_at=utc_now(),
            error=str(exc),
        )
        raise

    if args.answer:
        answer = extract_assistant_answer(lines)
        if not answer:
            answer = "".join(lines).strip()
        args.answer.write_text(answer + ("\n" if answer else ""), encoding="utf-8")

    cancelled = task_cancel_requested(args.task_state)
    if cancelled:
        status = "cancelled"
        error = None
    elif rc == 0:
        status = "completed"
        error = None
    else:
        status = "failed"
        error = f"Codex exited with status {rc}"
    patch_task(
        args.task_state,
        status=status,
        pid=None,
        finished_at=utc_now(),
        error=error,
    )
    return rc


if __name__ == "__main__":
    sys.exit(main())

File: tools/codex-scheduler/src/JobRunner.cpp
#include "Scheduler.h"
#include "Common.h"
#include "TaskStore.h"

#include <QtCore/QDir>
#include <QtCore/QFile>
#include <QtCore/QFileInfo>
#include <QtCore/QProcess>
#include <QtCore/QSaveFile>
#include <QtCore/QStandardPaths>
#include <QtCore/QTextStream>

namespace harr {

static std::optional<TaskStore> storeForTaskState(const QString &taskState)
{
    if (taskState.isEmpty()) return std::nullopt;
    QDir dir = QFileInfo(taskState).absoluteDir();
    if (dir.dirName() != "tasks") return std::nullopt;
    if (!dir.cdUp()) return std::nullopt;
    return TaskStore(dir.absolutePath());
}

int runJobRunner(const QStringList &args)
{
    QString logPath, answerPath, taskState;
    QStringList command;
    bool commandMode = false;
    for (int i = 0; i < args.size(); ++i) {
        const QString arg = args[i];
        if (commandMode) {
            command << arg;
            continue;
        }
        if (arg == "--") {
            commandMode = true;
            continue;
        }
        if (arg == "--log" && i + 1 < args.size()) logPath = args[++i];
        else if (arg == "--answer" && i + 1 < args.size()) answerPath = args[++i];
        else if (arg == "--task-state" && i + 1 < args.size()) taskState = args[++i];
        else {
            QTextStream(stderr) << "job-runner: unknown argument: " << arg << '\n';
            return 2;
        }
    }
    if (logPath.isEmpty() || command.isEmpty()) {
        QTextStream(stderr) << "job-runner: --log and command after -- are required\n";
        return 2;
    }

    QDir().mkpath(QFileInfo(logPath).absolutePath());
    QFile log(logPath);
    if (!log.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
        QTextStream(stderr) << "job-runner: cannot open log: " << log.errorString() << '\n';
        return 2;
    }

    auto taskStore = storeForTaskState(taskState);
    const QString taskId = taskState.isEmpty() ? QString{} : QFileInfo(taskState).completeBaseName();
    auto patchState = [&](const QJsonObject &patch) {
        if (!taskStore || taskId.isEmpty()) return;
        try { taskStore->patch(taskId, patch); } catch (...) {}
    };

    QString program = command.takeFirst();
    QStringList processArgs = command;
    const QString setsid = QStandardPaths::findExecutable("setsid");
    if (!setsid.isEmpty()) {
        processArgs.prepend(program);
        program = setsid;
    }

    QProcess process;
    process.setProcessChannelMode(QProcess::MergedChannels);
    process.start(program, processArgs);
    if (!process.waitForStarted(5000)) {
        patchState({{"status", "failed"}, {"pid", QJsonValue::Null},
                    {"finished_at", utcNow()}, {"error", process.errorString()}});
        QTextStream(stderr) << "job-runner: cannot start command: " << process.errorString() << '\n';
        return 127;
    }

    const qint64 pid = process.processId();
    patchState({{"status", "running"}, {"pid", pid}, {"started_at", utcNow()},
                {"finished_at", QJsonValue::Null}, {"error", QJsonValue::Null}});

    QByteArray pending;
    QByteArray rawAll;
    QStringList assistantMessages;
    auto consume = [&](const QByteArray &chunk, bool final) {
        if (!chunk.isEmpty()) {
            log.write(chunk);
            log.flush();
            rawAll += chunk;
            pending += chunk;
        }
        while (true) {
            const qsizetype newline = pending.indexOf('\n');
            if (newline < 0) break;
            const QByteArray line = pending.left(newline);
            pending.remove(0, newline + 1);
            for (const auto &event : parseCodexJsonLine(line)) {
                if (event.kind == "assistant") assistantMessages << event.text;
            }
        }
        if (final && !pending.isEmpty()) {
            for (const auto &event : parseCodexJsonLine(pending)) {
                if (event.kind == "assistant") assistantMessages << event.text;
            }
            pending.clear();
        }
    };

    while (process.state() != QProcess::NotRunning) {
        process.waitForReadyRead(250);
        consume(process.readAll(), false);
    }
    process.waitForFinished(1000);
    consume(process.readAll(), true);
    const int rc = process.exitStatus() == QProcess::NormalExit ? process.exitCode() : 128;

    if (!answerPath.isEmpty()) {
        QString answer = assistantMessages.join("\n\n").trimmed();
        if (answer.isEmpty()) answer = QString::fromUtf8(rawAll).trimmed();
        QDir().mkpath(QFileInfo(answerPath).absolutePath());
        QSaveFile answerFile(answerPath);
        if (answerFile.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
            if (!answer.isEmpty()) answer += '\n';
            answerFile.write(answer.toUtf8());
            answerFile.commit();
        }
    }

    bool cancelled = false;
    if (!taskState.isEmpty()) {
        if (auto state = readJsonObject(taskState)) cancelled = state->value("cancel_requested").toBool(false);
    }
    const QString finalStatus = cancelled ? "cancelled" : (rc == 0 ? "completed" : "failed");
    QJsonObject finalPatch{{"status", finalStatus}, {"pid", QJsonValue::Null}, {"finished_at", utcNow()}};
    if (!cancelled && rc != 0) finalPatch.insert("error", QString("Codex exited with status %1").arg(rc));
    else finalPatch.insert("error", QJsonValue::Null);
    patchState(finalPatch);
    return rc;
}

} // namespace harr
